features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/propagation/DatadogHttpCodec.java

```diff
@@ -6,116 +6,167 @@
-            traceId = validateUInt64BitsID(value, 10);
-          } else if (SPAN_ID_KEY.equalsIgnoreCase(key)) {
-            spanId = validateUInt64BitsID(value, 10);
-          } else if (SAMPLING_PRIORITY_KEY.equalsIgnoreCase(key)) {
-            samplingPriority = Integer.parseInt(value);
-          } else if (ORIGIN_KEY.equalsIgnoreCase(key)) {
-            origin = value;
-          } else if (key.startsWith(OT_BAGGAGE_PREFIX)) {
-            if (baggage.isEmpty()) {
-              baggage = new HashMap<>();
+            for (final Map.Entry<String, String> entry : context.baggageItems()) {
+                carrier.put(OT_BAGGAGE_PREFIX + entry.getKey(), HttpCodec.encode(entry.getValue()));
             }
-            baggage.put(key.replace(OT_BAGGAGE_PREFIX, ""), HttpCodec.decode(value));
-          }
+            // adding the tags
+            carrier.put(DATADOG_TAGS_KEY, MOST_SIGNIFICANT_TRACE_ID_KEY + "=" + mostSignificantTraceId);
+
+
+            // always use max sampling priority for Android traces
+            carrier.put(SAMPLING_PRIORITY_KEY, "1");
+        }
+    }
 
-          if (taggedHeaders.containsKey(key)) {
-            if (tags.isEmpty()) {
-              tags = new HashMap<>();
+    public static class Extractor implements HttpCodec.Extractor {
+        private final Map<String, String> taggedHeaders;
+
+        public Extractor(final Map<String, String> taggedHeaders) {
+            this.taggedHeaders = new HashMap<>();
+            for (final Map.Entry<String, String> mapping : taggedHeaders.entrySet()) {
+                this.taggedHeaders.put(mapping.getKey().trim().toLowerCase(Locale.US), mapping.getValue());
             }
-            tags.put(taggedHeaders.get(key), HttpCodec.decode(value));
-          }
         }
 
-        if (!BigInteger.ZERO.equals(traceId)) {
-          final ExtractedContext context =
-              new ExtractedContext(traceId, spanId, samplingPriority, origin, baggage, tags);
-          context.lockSamplingPriority();
+        @Override
+        public SpanContext extract(final TextMapExtract carrier) {
+            try {
+                Map<String, String> baggage = Collections.emptyMap();
+                Map<String, String> tags = Collections.emptyMap();
+                BigInteger spanId = BigInteger.ZERO;
+                int samplingPriority = PrioritySampling.UNSET;
+                String origin = null;
+                String mostSignificant64BitsTraceIdAsHex = null;
+                String leastSignificant64BitsTraceIdAsDecimal = null;
+
+
+                for (final Map.Entry<String, String> entry : carrier) {
+                    final String key = entry.getKey().toLowerCase(Locale.US);
+                    final String value = entry.getValue();
+
+                    if (value == null) {
+                        continue;
+                    }
+
+                    if (LEAST_SIGNIFICANT_TRACE_ID_KEY.equalsIgnoreCase(key)) {
```

Comment:
nit: do we need ignorecase when the constant is lowercase and we're doing toLowerCase on the key? same question for some others here

Reply:
didn't touch that code, was just refactoring of the format